Create s3-private module and pvsite ml bucket

devsjc opened the pull request:
Genericise s3 module and update deprecated resources

peterdudfield commented on terraform/modules/storage/s3-private/variables.tf:
```diff
@@ -0,0 +1,24 @@
+
+variable "lifecycled_prefixes" {
+  description = "The data-slash prefixes requiring lifecycles"
+  type = list(string)
```
can you have the option of none? If we want to keep some buckets all the time

devsjc replied:
Yes, an empty list does that [], like in the case for the s3-ml one in main

devsjc replied:
Also, I think this only setting lifecycles on prefixes within the bucket, as opposed to on the entire bucket

peterdudfield commented on terraform/modules/storage/s3-private/variables.tf:
```diff
@@ -0,0 +1,24 @@
+  description = "The name of the service for which the bucket supplies"
+  type = string
+}
+
```
do we have an option of how long each bucket will keeps its data?

devsjc commented on terraform/modules/storage/s3-private/bucket.tf:
```diff
@@ -0,0 +1,49 @@
+      prefix = "${each.key}/"
+    }
+    expiration {
+      days = 7
```
It only creates these 7 day lifecycles for the specified prefixes - this mirrors the setup prior to this refactor: 

```
# Documentation: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket
resource "aws_s3_bucket" "s3-nwp-bucket" {
  bucket = "nowcasting-nwp-${var.environment}"
  acl    = "private"

  lifecycle_rule {
    id      = "remove_old_files"
    enabled = true

    prefix = "data/"

    expiration {
      days = 7
    }
  }

  lifecycle_rule {
    id      = "remove_old_raw_files"
    enabled = true

    prefix = "raw/"

    expiration {
      days = 7
    }
  }

  tags = {
    Name        = "${var.environment}-s3"
    Environment = "${var.environment}"
  }
}
```

devsjc commented:
I have removed all changes to the nowcasting side of things, moved the changes into their own module that PVSite uses to create its ml bucket, left s3 to call the old module so no buckets will be deleted over there

peterdudfield commented:
this should do this one - https://github.com/openclimatefix/ocf-infrastructure/issues/155?
